src/controllers/blocks/BlocksController.ts

```diff
@@ -94,14 +95,26 @@ export default class BlocksController extends AbstractController<BlocksService>
 		const eventDocsArg = eventDocs === 'true';
 		const extrsinsicDocsArg = extrinsicDocs === 'true';
 
-		const hash =
-			finalized === 'false' || !this.options.finalizes
-				? (await this.api.rpc.chain.getHeader()).hash
-				: await this.api.rpc.chain.getFinalizedHead();
+		let hash;
+		let queryFinalizedHead;
+
+		if (finalized === 'false' || !this.options.finalizes) {
```

Comment:
As far as I know express enforces the type on a query as `string | QueryString.ParsedQs`, but I think this is a great question. I believe if we were to enforce a type annotation on the query we would need to write a converter on the queryString itself, then enforce the type annotation after. @emostov Any thoughts on this.

Reply:
Yeah, express actually has quite a few caveats with the query params. It can be a number of different data types, so making an assumption that it will just be a string representing a boolean we can cast can be semi dangerous - it likely won't fail at runtime since it will either be a truthy or a falsey value but I think it can make things harder to debug if we make assumptions. Checkout https://evanhahn.com/gotchas-with-express-query-parsing-and-how-to-avoid-them/

Reply:
That was an interesting read, JS is still scary after all these years. 
I'm not sure we're talking about the same thing though so re-stating it explicitly (pardon the likley wrong TS syntax):

Can't we change line 92, `{ query: { eventDocs, extrinsicDocs, finalized } }`, into something like this:

```
{ query: { eventDocs: boolean, extrinsicDocs: boolean, finalized: boolean } }
```

i.e. not allow "truthy stuff" at all, but only ever accept bona fide `boolean`s? I realize this would require changes in the query string parsing – adding a middleware? configure a strict query string parser? – but wouldn't that be a fairly big win to us overall?

From reading the article above (TIL!) it seems like Express took a user friendly route by providing a "smart" query parser. We're an API-only app though and being strict about query parameters is a feature and not a hindrance to us. If we did make such a change and could farm out the query parsing we'd be able to clean up endpoints such as this one and also sleep better at night knowing things are locked down by default (as opposed to flexible by default).
This is not something we need to jump on right away (and definitely not in this PR) but I think it's worth surveying the space and see if there is a query parser out there that is a) well maintained and with a good reputation, and that b) plays nice with TS type annotations.

Reply:
Yea I agree that if we implement a middleware to almost sanitize the query types it would be a nice improvement. I think it's definitely something to benefit from when it comes to an extra layer of definitively accurate return values. Definitely something for a future PR, but also something to put some good thought into, and maybe have a chat about in the near future to design it. 

In regards to the code snippet above we cant actually do: 

```
{ query: { eventDocs: boolean, extrinsicDocs: boolean, finalized: boolean } }
```

Reason being, when we destructure an object like above, javascript thinks we are assigning a key `eventDocs` to a value `boolean` ie: a variable like `const boolean = false` which is not declared anywhere. 

If we were to add types on something like this it would look like:

```
Interface RequestHandlerExample {
  query: QueryHandler
}

Interface QueryHandler {
  eventDocs: string,
  extrinsicDocs: string,
  finalized: string 
}

{ query: { eventDocs, extrinsicDocs, finalized } }: RequestHandlerExample
```

That also being said the types for the query are already set by the RequestHandler for the express request and response so we can't/shouldn't override them necessarily.

Reply:
We can try out some middleware. Not sure what is out there off the shelf, but I do remember trying to do something like this with query params. From what I recall the best that was possible is rejecting a non-boolean query param, and then in the controller you can assumed its `undefined | boolean`. The reason being is that you can't change data in the express `Request` object as it goes through the middleware pipeline (if we could we could force `boolean | undefined`). Its been awhile though so I could be wrong. Will make an issue

Reply:
Issue bool query param validation https://github.com/paritytech/substrate-api-sidecar/issues/398

Reply:
Ty @TarikGul for a better TS example. :)

We could definitely have a whiteboard session about this, it'd be fun actually. What I'm after is something a wee bit different than #398 (but it is starting to sound like what I'd ideally want is really hard/impossible). What I'd want is a way to make the TS compiler enforce type validation at compile time, e.g. by injecting endpoint specific middleware or by injecting code into the endpoint function directly. E.g. at compile time, look at the function signature and find too `bool`s and an array of `String`, inject middleware that rejects all function calls that don't respect the contract. But yeah, it doesn't seem like something that is possible today with TS.

Reply:
@dvdplm sounds like you want macros? Not sure what meta-programming features like this exist in TS/JS land